test/powershell/engine/Api/MultithreadedEngine.Tests.ps1

```diff
@@ -0,0 +1,198 @@
+            foreach ($i in 1..20) {
+                $d1.Add($foreach.Current)
+                $foreach.MoveNext() > $null
+                $d2.Add($foreach.Current)
+            }
+            $d1.Complete()
+            $d2.Complete()
+            $script = [scriptblock]::Create("`$input${sbStub}")
+        }
+
+        It 'can invoke multiple scripts asynchronously with input' {
+            $r1 = [powershell]::Create().AddScript($script).InvokeAsync($d1)
+            $r2 = [powershell]::Create().AddScript($script).InvokeAsync($d2)
+            [System.Threading.Tasks.Task]::WaitAll(@($r1, $r2))
+            $r1.Status | Should -Be ([System.Threading.Tasks.TaskStatus]::RanToCompletion)
+            $r1.IsCompletedSuccessfully | Should -Be $true
+            $r2.Status | Should -Be ([System.Threading.Tasks.TaskStatus]::RanToCompletion)
+            $r2.IsCompletedSuccessfully | Should -Be $true
+            $sortedResults = @($r1.Result) + @($r2.Result) | Sort-Object -Property Time
+            Compare-Object -ReferenceObject @(1..20) -DifferenceObject $sortedResults.Value -SyncWindow 20 | Should -Be $null
+        }
+
+        It 'can invoke multiple scripts asynchronously with input and capture output' {
+            $o = New-Object -TypeName 'System.Management.Automation.PSDataCollection[PSObject]'
+            $r1 = [powershell]::Create().AddScript($script).InvokeAsync($d1, $o)
+            $r2 = [powershell]::Create().AddScript($script).InvokeAsync($d2, $o)
+            [System.Threading.Tasks.Task]::WaitAll(@($r1, $r2))
+            $o.Complete()
+            $r1.Status | Should -Be ([System.Threading.Tasks.TaskStatus]::RanToCompletion)
+            $r1.IsCompletedSuccessfully | Should -Be $true
+            $r2.Status | Should -Be ([System.Threading.Tasks.TaskStatus]::RanToCompletion)
+            $r2.IsCompletedSuccessfully | Should -Be $true
+            $sortedResults = $o | Sort-Object -Property Time
+            Compare-Object -ReferenceObject @(1..20) -DifferenceObject $sortedResults.Value -SyncWindow 20 | Should -Be $null
+        }
+    }
+
+    Context 'PowerShell::StopAsync' {
+        It 'can stop multiple scripts that are running asynchronously' {
+            $sb1 = [scriptblock]::Create("@(1,3,5,7,9,11,13,15,17,19)${sbStub}")
+            $sb2 = [scriptblock]::Create("@(2,4,6,8,10,12,14,16,18,20)${sbStub}")
+            $ps1 = [powershell]::Create().AddScript($sb1)
+            $ps2 = [powershell]::Create().AddScript($sb2)
+            $ps1.InvokeAsync() > $null
+            $ps2.InvokeAsync() > $null
+            while ($ps1.InvocationStateInfo.State -ne [System.Management.Automation.PSInvocationState]::Running -and
+                   $ps2.InvocationStateInfo.State -ne [System.Management.Automation.PSInvocationState]::Running) {
+                Start-Sleep -Milliseconds 100
+            }
+            $sr1 = $ps1.StopAsync({}, $null)
+            $sr2 = $ps2.StopAsync({}, $null)
```

Comment:
After `StopAsync`, I would also like a test for the state of the task objects return from the previous `InvokeAsync` calls.

Reply:
Added.